### ISSUE TITLE
macos-latest cannot install homebrew packages after using this action

### ISSUE THREAD

**Issue by ReenigneArcher** [github/codeql-action#2266]
macos-latest (i.e. macos-14) image fails to install homebrew packages AFTER using codeql-action. The same homebrew install works fine if not using the codeql-action.

I filed a report on the runner-images repo (before I realized the codeql action was causing the issue), they suggested I file a bug report here instead.
https://github.com/actions/runner-images/issues/9760

This is the simplest reproduction I could create. https://github.com/ReenigneArcher/brew-error-reproduction/blob/master/.github/workflows/brew_error.yml

It works fine if not using the codeql setup step first.

Logs: https://github.com/ReenigneArcher/brew-error-reproduction/actions/runs/8870403949/job/24352158742

**Comment by aibaars** [comment]
Thanks for reporting! Could you re-run the failed job with [debug logging enabled](https://github.blog/changelog/2022-05-24-github-actions-re-run-jobs-with-debug-logging/)? When running in debug mode, the CodeQL Action should attach a debug artifact to the Job's summary page. That would be very helpful for diagnosing the problem.

As a workaround, have you tried running "brew install" before the `github/codeql-action/init` step?

**Comment by ReenigneArcher** [comment]
I have re-run it in debug mode: https://github.com/ReenigneArcher/brew-error-reproduction/actions/runs/8870403949/job/24376380948

I have not tried refactoring the workflow yet, but I know that the brew install would at least succeed if it were before the init. The actual production workflow is a bit more complex than the reproduction workflow, I'd like to avoid refactoring it, if possible.

**Comment by aibaars** [comment]
> The actual production workflow is a bit more complex than the reproduction workflow, I'd like to avoid refactoring it, if possible.

Another workaround you could try is to change the `runs-on:` property of the workflow. You could try switching to an older version of macos or to an intel-based macos runner. It's not great, but might be easier an easier tweak than re-factoring the workflow.

**Comment by ReenigneArcher** [comment]
Re-ordering the workflow won't be feasible due to it requiring the build to run after this action is initialized: https://github.com/orgs/community/discussions/68044#discussioncomment-7109477

Mostly this is due to me using a single code-ql workflow for all my org repos, which checks for the existence of a `.codeql-prebuild-<language>-<runner_OS>.sh` file... and I'd rather not have multiple scripts for code-ql for each os... e.g. deps and build.

**Comment by ReenigneArcher** [comment]
Just checking in. This is still occurring: https://github.com/ReenigneArcher/brew-error-reproduction/actions/runs/9916092767/job/27397917114#step:3:66

**Comment by angelapwen** [comment]
Thank you for sharing your workflow runs — this looks like a new issue with our build tracer. We've raised this with the appropriate team internally.

**Comment by angelapwen** [comment]
After a colleague dug into this a bit more, here's what we've come up with: our build tracer forces relocation to x86, which leads to a x86 execution under Rosetta. Apparently, that x86 execution doesn't play well with Homebrew. For now, please try the suggestion in the logs, `arch -arm64 brew install ...` to force Homebrew back to the arm64 execution.

**Comment by ReenigneArcher** [comment]
@angelapwen Thank you for looking into this. I am pretty sure I tried the suggestion in the logs (`arch -arm64 brew install ...`) previously and it did not help... I will try again though and report back.

**Comment by ReenigneArcher** [comment]
Sorry for the delay... When using the suggestion from the logs (https://github.com/ReenigneArcher/brew-error-reproduction/pull/2), I get the following error:

```
arch: posix_spawnp: brew: Bad CPU type in executable
```
https://github.com/ReenigneArcher/brew-error-reproduction/actions/runs/11108869334/job/30862839806?pr=2#step:3:48

Interesting since homebrew suggests to use that exact command.

**Comment by ReenigneArcher** [comment]
Just following up to hopefully keep this on the radar.

The macos-12 image is about to removed. macos-13 will likely be EOL in October 2025, so there's still a little time.

These also seem related:

- https://github.com/github/codeql-action/issues/2347
- https://github.com/github/codeql-action/issues/2109

**Comment by smowton** [comment]
This should be fixed in CodeQL 2.19.4: an `arch -arm64` command would often fail because CodeQL would create and try to invoke an x86-64-only binary with the explicit architecture flag passed to `posix_spawn` by `arch`. Now we suppress that behaviour when `posix_spawn` forces use of `arm64`, though it still is not possible to trace compiler invocations reached this way, so this is useful for `arch -arm64 brew install`, but not `arch -arm64 gcc analyse_this.c -c -o ...` for example.

**Comment by ReenigneArcher** [comment]
@smowton when will this action start using 2.19.4? It looks like it's still using 2.19.3.

**Comment by smowton** [comment]
It became fully enabled by default at around 4.30pm on the 3rd Dec; try a rerun now?

**Comment by ReenigneArcher** [comment]
Yes, looks like it's using the new version now. Seems to work if forcing arm for brew, but not using the normal brew commands: https://github.com/ReenigneArcher/brew-error-reproduction/actions/runs/12160656532/job/33913434744#step:3:62

Good enough though, thank you!

**Comment by ReenigneArcher** [comment]
@smowton sorry for the ping, but I'm facing an issue with this still. Just got around to updating the runners to `macos-latest` from `macos-13`. Two of three repos everything went fine, but for one I am having issues linking.

`ld: symbol(s) not found for architecture x86_64`: https://github.com/LizardByte/Sunshine/actions/runs/13906824811/job/38911761299?pr=3724#step:6:1093

For some reason clang seems to be looking for symbols of `x86_64` instead of `arm64`. Any idea how to solve this?

This is from PR: https://github.com/LizardByte/Sunshine/pull/3724

**Comment by smowton** [comment]
So, we're now getting through the `brew install` command but failing during final linkage. This is because the libraries installed are arm64-only, but CodeQL is forcing something in the build pipeline to run in x86-64 mode.

This is because we can't trace `arm64e` binaries (most MacOS system binaries are arm64e, not plain arm64). Therefore when a (x86-64, arm64e) binary is executed, we force execution of the x86 slice, which in turn means its children are executed as x86 too. In this case that leads to the link step attempting to target x86 with only ARM libraries available.

Possible workarounds:
* Do you need a MacOS build at all here? For security analysis purposes, is there any significant deviation between code compiled on different platforms?
* If you really do, consider setting up and running an x86 build on MacOS, to avoid these platform-switching issues -- that would likely mean installing x86 Homebrew and pulling the x86 versions of your dependencies.
* Alternatively, can the build be explicitly told its target platform rather than guess it based on the environment? If the x86 tooling can produce ARM binaries, that could work.

**Comment by ReenigneArcher** [comment]
> If you really do, consider setting up and running an x86 build on MacOS, to avoid these platform-switching issues -- that would likely mean installing x86 Homebrew and pulling the x86 versions of your dependencies.

This worked, although the build took > 1 hrs, versus the standard ~15 minutes.

Basically just had to add this to my `prebuild.sh` script.

```bash
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
eval "$(/usr/local/bin/brew shellenv)"
```
